DTLS: Add server side stateless and CID QoL API

julek-wolfssl opened the pull request:
- wolfDTLS_accept_stateless - statelessly listen for incoming connections
- wolfSSL_inject - insert data into WOLFSSL object
- wolfSSL_SSL(Enable|Disable)Read - enable/disable reading from IO
- wolfSSL_get_wfd - get the write side file descriptor
- wolfSSL_dtls_set_pending_peer - set the pending peer that will be upgraded to regular peer when we successfully de-protect a DTLS record
- wolfSSL_dtls_get0_peer - zero copy access to the peer address
- wolfSSL_is_stateful - boolean to check if we have entered stateful processing
- wolfSSL_dtls_cid_get0_rx - zero copy access to the rx cid
- wolfSSL_dtls_cid_get0_tx - zero copy access to the tx cid
- wolfSSL_dtls_cid_parse - extract cid from a datagram/message

julek-wolfssl committed:
DTLS: Add server side stateless and CID QoL API

- wolfDTLS_accept_stateless - statelessly listen for incoming connections
- wolfSSL_inject - insert data into WOLFSSL object
- wolfSSL_SSL(Enable|Disable)Read - enable/disable reading from IO
- wolfSSL_get_wfd - get the write side file descriptor
- wolfSSL_dtls_set_pending_peer - set the pending peer that will be upgraded to regular peer when we successfully de-protect a DTLS record
- wolfSSL_dtls_get0_peer - zero copy access to the peer address
- wolfSSL_is_stateful - boolean to check if we have entered stateful processing
- wolfSSL_dtls_cid_get0_rx - zero copy access to the rx cid
- wolfSSL_dtls_cid_get0_tx - zero copy access to the tx cid
- wolfSSL_dtls_cid_parse - extract cid from a datagram/message

rizlik reviewed (commented):
I think this is a big step in the right direction but there are things unclear. These comments are considering https://github.com/wolfSSL/wolfssl-examples/pull/472 as the user of the changes in this PR.

1. what's the purpose of splitting the fd in write and read? 

2. I don't get why we should use the pendingPeer and the example doesn't use it. It looks complicate and there is code that get hit after we copy the pendingPeer  to the current Peer that it doesn't look doing nothing other than flipping the processPendingRecord bit. 

3. I don't see much gain in using wolfSSL_accept_stateless. 
AFAIU it just boils down in a) disabling read so that it can return before reading from the socket and b) to return after a good CH. a) avoids dropping other peer messages that can be received in the EmbedRecvFrom. But, as shown in the example, the server has already to read the messages itself outside of the wolfSSL to do proper demux. I wonder if then this is superfluous.  

I would suggest of using a custom internal field like the pendingPeer, but called, as you suggested, currentRecordPeer.
Until the ssl is stateless, this is set on recvfrom, but no filtering happens bc of this. 
It's used as the destination of the HRR/HVR and to compute the cookie. 
It's not used anywhere else and it's not automatically copied over the other peer.
This way an ssl object can be used to reply to the CHs w/o cookies.
If we got a CH w/ valid cookie we can then leverage our existing goodch callback to set the right peer for sending and continue the connection.
If the application wants to demux, it needs to manage the reading from the socket itself anyway.

Just to recap, my suggestion are:
- In EmbedRecvFrom to set a "peer to send the HRR/HRV" to if we are in a stateless
- Application needs to use ChGoodCB to set the proper peer (is it already like this?)
- Make clear that application needs to manage the read itself if it wants to demux over a single socket

The new APIs (inject, parse_cid) allows the application to manage the connection with a good tradeoff of complexity and control imo.

rizlik commented on doc/dox_comments/header_files/ssl.h:
```diff
@@ -3899,6 +4010,40 @@ int  wolfSSL_dtls_set_peer(WOLFSSL* ssl, void* peer, unsigned int peerSz);
+    \sa wolfSSL_dtls_set_peer
+    \sa wolfSSL_dtls
+*/
+int  wolfSSL_dtls_get0_peer(WOLFSSL* ssl, const void** peer, unsigned int* peerSz);
```
don't expose the internal cid.
Prefer to add an API to compare to the ssl cid object if afraid of performance loss of copying the cid.
(like `wolfSSL_dtls_cid_cmp(WOLFSSL*ssl,byte *cid, int cidSz)`)

julek-wolfssl replied:
This is for the peer address. Did you mean `wolfSSL_dtls_cid_get0_rx`? I think its better to allow the user to compare the CID however they want (using many `memcmp` or a bloom filter for example).

rizlik replied:
I do think it exposes our internal implementation, if we haven't any request of doing that I'll avoid it. but I'm ok if you think it's important.

julek-wolfssl replied:
It only exposes the peer directly. The parameter is marked `const` so it signals to the user that this should not be modified.

julek-wolfssl committed:
Client TLS: Set traffic decrypt keys when parsing Finished

julek-wolfssl commented on doc/dox_comments/header_files/ssl.h:
```diff
@@ -14138,6 +14283,35 @@ int  wolfSSL_write_early_data(WOLFSSL* ssl, const void* data,
+/*!
+    \ingroup IO
+
+    \brief
```
Adding.

julek-wolfssl committed:
Code review and jenkins fixes

julek-wolfssl commented on doc/dox_comments/header_files/ssl.h:
```diff
@@ -14967,6 +15142,27 @@ int wolfSSL_dtls_cid_get_rx(WOLFSSL* ssl, unsigned char* buffer,
+\brief Get the ConnectionID used by the other peer. See RFC 9146 and RFC
+9147.
+
+ \return WOLFSSL_SUCCESS if ConnectionID was correctly copied, error code
```
Fixing.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -1372,10 +1397,40 @@ int wolfSSL_dtls_cid_get_tx(WOLFSSL* ssl, unsigned char* buf,
+        return;
+    if (msg[0] == dtls12_cid) {
+        /* DTLS 1.2 CID packet */
+        if (msgSz < DTLS_RECORD_HEADER_SZ + cidSz)
+            return;
+        /* content type(1) + version(2) + epoch(2) + sequence(6) */
+        *cid = msg + ENUM_LEN + VERSION_SZ + OPAQUE16_LEN + OPAQUE16_LEN +
+                OPAQUE32_LEN;
```
I'm not sure what you mean?

rizlik replied:
we are checking `DTLS_RECORD_HEADER_SZ + cidSz` but then we are coping from offset `ENUM_LEN, VERSION_SZ, LEN, LEN and OPAQUE32_LEN`

julek-wolfssl replied:
The CID field is not the last field which is why the checks are different. `DTLS_RECORD_HEADER_SZ` is the sum of the remaining static fields.
```
     struct {
         ContentType outer_type = tls12_cid;
         ProtocolVersion version;
         uint16 epoch;
         uint48 sequence_number;
         opaque cid[cid_length];               // New field
         uint16 length;
         opaque enc_content[DTLSCiphertext.length];
     } DTLSCiphertext;
````

julek-wolfssl commented on src/ssl.c:
```diff
@@ -9202,15 +9321,21 @@ int wolfSSL_dtls_retransmit(WOLFSSL* ssl)
-        int result = DtlsMsgPoolSend(ssl, 0);
+        int result;
+#ifdef WOLFSSL_DTLS13
+        if (IsAtLeastTLSv1_3(ssl->version))
```
General improvement. Adding 1.3 support to `wolfSSL_dtls_retransmit`.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -1855,45 +1866,97 @@ int wolfSSL_dtls_free_peer(void* addr)
+
+    if (ssl->buffers.dtlsCtx.peer.sa != NULL &&
+            ssl->buffers.dtlsCtx.peer.sz == peerSz &&
+            XMEMCMP(ssl->buffers.dtlsCtx.peer.sa, peer, peerSz) == 0) {
```
Using `sockAddrEqual`.

julek-wolfssl commented on src/internal.c:
```diff
@@ -11520,15 +11525,15 @@ static int GetDtlsRecordHeader(WOLFSSL* ssl, word32* inOutIdx,
             return LENGTH_ERROR;
-        if (cidSz > DTLS_CID_MAX_SIZE ||
-                wolfSSL_dtls_cid_get_rx(ssl, cid, cidSz) != WOLFSSL_SUCCESS)
+        if (cidSz != DtlsGetCidRxSize(ssl) ||
```
You're right. Removing.

julek-wolfssl commented on src/ssl.c:
```diff
@@ -2874,6 +2958,42 @@ int wolfSSL_write(WOLFSSL* ssl, const void* data, int sz)
+
+    WOLFSSL_ENTER("wolfSSL_read_internal");
+
+    if (ssl == NULL || data == NULL || sz <= 0)
```
I don't see how a 0-length inject would be valid.

julek-wolfssl commented on src/dtls.c:
```diff
@@ -1372,10 +1406,40 @@ int wolfSSL_dtls_cid_get_tx(WOLFSSL* ssl, unsigned char* buf,
     return DTLS_CID_MAX_SIZE;
 }
+
+void wolfSSL_dtls_cid_parse(const unsigned char* msg, unsigned int msgSz,
```
`cid` is already the return. I don't see the point of complicating with an extra return. If anything, I can have the function return the pointer instead of outputting it to a parameter.

rizlik replied:
ok, I do think checking for the return is more uniform than checking if *cid is NULL. Please add this to the documentation of the API.

julek-wolfssl replied:
Ok so I'll change it to return the CID instead of setting the parameter.

julek-wolfssl commented:
> what's the purpose of splitting the fd in write and read?

It was always split. I just added a getter for that field.

> I don't get why we should use the pendingPeer and the example doesn't use it. It looks complicate and there is code that get hit after we copy the pendingPeer to the current Peer that it doesn't look doing nothing other than flipping the processPendingRecord bit. 
[...]
 It's used as the destination of the HRR/HVR and to compute the cookie.

The pending peer is used in the example when CID's are used. The bit flipping is just to make sure that the pending peer is only valid for the next parsed record. We already have a good mechanism for cookie calculation, there is no need to change it. The pendingPeer is used in conjunction with CID's to validate the incoming record before updating the peer address. An off-path attacker sending records with the CID set can change the address if we don't first attempt to de-protect the record. Attempting to de-protect an off-path attacker's packets should fail and we will reject the pendingPeer internally without any extra help from the application needed.

> I don't see much gain in using wolfSSL_accept_stateless.

You're the one who suggested adding it :D. I think it makes sense to have it to abstract away from the user the complexity of the stateless handling.

> But, as shown in the example, the server has already to read the messages itself outside of the wolfSSL to do proper demux. I wonder if then this is superfluous.

But we also support a server object that reads from the network directly. We have to support that use case as well. This API helps a lot with that case to make sure no records are lost.

To respond to the suggestions:

> In EmbedRecvFrom to set a "peer to send the HRR/HRV" to if we are in a stateless

Adding.

> Application needs to use ChGoodCB to set the proper peer (is it already like this?)

It does not need to do that. It calls `wolfSSL_dtls_set_peer` for peers during the connection process. `wolfSSL_dtls_set_pending_peer` is used during the connection with CID's.

> Make clear that application needs to manage the read itself if it wants to demux over a single socket

I think that is clear from the example. Where do you think we need more documentation?

rizlik commented on src/internal.c:
```diff
@@ -21533,6 +21550,23 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
+                    }
+                }
+            }
+#endif
```
this and the other block to update the peer feel misplaced here. One reason is that it will run for every packet, even the ones where we don't have set any pending peer.  Another reason is that is adding more noise in the already quite convoluted processReply. 

But I can't find a good solution because I don't see a good point where we can clear the pending peer on error, as we have many exit points in this function.

The only idea I've is to modify `wolfssl_inject` so that it also invokes `wolfSSL_negotiate`  if necessary and `wolfSSL_read`. If there are application-data we can return them from `wolfssl_inject`.

The basic idea is that we include what the app is doing inside `dispatchExistingConnection` inside the `wolfSSl_inject` API.

If this makes sens to you  we can include the the peer in the  api itself. It's true what we are doing two things in a single API, but it's also true that the data that we want to inject usually have a source, so the two information are related.

rizlik replied:
If you reject the `inject` modification, can you move this and the other block in separate functions and adds a longer comments on why we are doing that?

julek-wolfssl replied:
> run for every packet

Only bit flipping will run for every class.

> But I can't find a good solution because I don't see a good point where we can clear the pending peer on error, as we have many exit points in this function.

That is precisely the reason for this. If we could have a single `cleanup` goto it would make it easier but this function is complicated enough.

I don't want to add anything else to the inject API. I think it should do only what the user of an inject API would expect. I'll add more documentation and refactor it to make it cleaner.

julek-wolfssl committed:
Refactor dtls pending peer processing

rizlik reviewed (changes requested):
This looks better, thanks!

julek-wolfssl committed:
Protect peer access when WOLFSSL_RW_THREADED

julek-wolfssl commented on src/internal.c:
```diff
@@ -21341,16 +21342,58 @@ static int removeMsgInnerPadding(WOLFSSL* ssl)
+            /* Pending peer present and record deprotected. Update the peer. */
+            dtlsClearPeer(&ssl->buffers.dtlsCtx.peer);
+            ssl->buffers.dtlsCtx.peer = ssl->buffers.dtlsCtx.pendingPeer;
+            XMEMSET(&ssl->buffers.dtlsCtx.pendingPeer, 0,
```
Added rwlock.

julek-wolfssl committed:
wolfSSLReceive: Error return on interrupted connection

Interrupted connection should return control to the user since they may want to handle the signal that caused the interrupt. Otherwise, we might never give back control to the user (the timeout would error out but that causes a big delay).

socat.yml: in test 475, the test would send a SIGTERM after 3 seconds. We would continue to ignore this signal and continue to call `recvfrom`. Instead we should error out and give control back to the user.

julek-wolfssl committed:
fixup! Code review and jenkins fixes

julek-wolfssl committed:
fixup! DTLS: Add server side stateless and CID QoL API

julek-wolfssl commented:
Retest this please

douzzer reviewed (changes requested):
We can't yet test this PR because there's ASAN trouble in `master` with `--enable-dtlscid`:
```
[sp-all-asm-sanitizer] [1 of 1] [836ee1cbd5]
    autogen.sh 836ee1cbd5...   real 0m16.583s  user 0m14.716s  sys 0m0.962s
    configure...   real 0m20.741s  user 0m8.985s  sys 0m13.738s
    build...   real 3m1.504s  user 11m4.632s  sys 0m17.800s
    check...   real 0m37.326s  user 0m32.207s  sys 0m22.919s
=================================================================
==21529==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd3c8c5c5e at pc 0x7f8796ec37f2 bp 0x7ffd3c8a58f0 sp 0x7ffd3c8a5098
READ of size 31 at 0x7ffd3c8c5c5e thread T0
    #0 0x7f8796ec37f1 in strcmp /tmp/portage/sys-devel/gcc-14.2.1_p20241116/work/gcc-14-20241116/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:470
    #1 0x55f16a76ef1b in test_dtls13_basic_connection_id tests/api.c:99636
    #2 0x55f16a7a9f2d in ApiTest tests/api.c:102498
    #3 0x55f16a12cb84 in unit_test tests/unit.c:219
    #4 0x7f8794043289 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #5 0x7f879404334a in __libc_start_main_impl ../csu/libc-start.c:360
    #6 0x55f16a12c654 in _start ../sysdeps/x86_64/start.S:115

Address 0x7ffd3c8c5c5e is located in stack of thread T0 at offset 131710 in frame
    #0 0x55f16a76b4af in test_dtls13_basic_connection_id tests/api.c:99543

  This frame has 9 object(s):
    [48, 56) 'ctx_c' (line 99579)
    [80, 88) 'ctx_s' (line 99579)
    [112, 120) 'ssl_c' (line 99580)
    [144, 152) 'ssl_s' (line 99580)
    [176, 216) 'params' (line 99550)
    [256, 131360) 'test_ctx' (line 99581)
    [131616, 131626) 'client_cid' (line 99547)
    [131648, 131658) 'server_cid' (line 99548)
    [131680, 131710) 'readBuf' (line 99549) <== Memory access at offset 131710 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /tmp/portage/sys-devel/gcc-14.2.1_p20241116/work/gcc-14-20241116/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:470 in strcmp
Shadow bytes around the buggy address:
  0x7ffd3c8c5980: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5a00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5a80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5b00: f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x7ffd3c8c5b80: f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
=>0x7ffd3c8c5c00: 00 02 f2 f2 00 02 f2 f2 00 00 00[06]f3 f3 f3 f3
  0x7ffd3c8c5c80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5d00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5d80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5e00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x7ffd3c8c5e80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
    sp-all-asm-sanitizer fail_analytic_check
    failed config: '--srcdir' '.' '--disable-jobserver' '--enable-option-checking=fatal' '--enable-all' '--enable-testcert' '--enable-acert' '--enable-dtls13' '--enable-dtls-mtu' '--enable-dtls-frag-ch' '--enable-dtlscid' '--enable-quic' '--with-sys-crypto-policy' '--enable-intelasm' '--enable-sp-asm' '--enable-sp-math-all' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK -pedantic' 'CC=gcc-14' 'LDFLAGS=-g -fno-omit-frame-pointer -fsanitize-recover=all -fsanitize=address,pointer-subtract,leak,undefined,float-cast-overflow,float-divide-by-zero,bounds-strict -fsanitize-recover=all ' 'CFLAGS=-DWC_DEBUG_CIPHER_LIFECYCLE -g -fno-omit-frame-pointer -fsanitize=address,pointer-subtract,leak,undefined,float-cast-overflow,float-divide-by-zero,bounds-strict -fsanitize-recover=all --param=max-vartrack-size=128000000 -march=native'
    BUILD_ENV: 'FAIL_BUILD_CODENAME=fail_analytic_build' 'MAX_FIPS_CODE_SZ=10000000'
    RUN_ENV: 'LD_LIBRARY_PATH=/tmp/tmp.4346_11739/wolfssl_test_workdir.15879/wolfssl/src/.libs:/usr/lib/gcc/x86_64-pc-linux-gnu/14:/usr/lib/gcc/x86_64-pc-linux-gnu/14/32' 'ASAN_OPTIONS=halt_on_error=0 color=always log_path=sanitizer_log.asan ' 'UBSAN_OPTIONS=halt_on_error=0 color=always log_path=sanitizer_log.ubsan ' 'LSAN_OPTIONS=halt_on_error=0 color=always log_path=sanitizer_log.lsan ' 'MSAN_OPTIONS=halt_on_error=0 color=always log_path=sanitizer_log.msan ' 'TSAN_OPTIONS=halt_on_error=0 color=always log_path=sanitizer_log.tsan '
```
Another problem I ran into during testing, also in `master`, is `clang-tidy` carps:
```
[clang-tidy-all-sp-all] [8 of 39] [556bd38c64]
    configure...   real 0m27.793s  user 0m12.559s  sys 0m17.234s
    build...90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2426)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/server/server.c:2426:21: warning: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119 [clang-analyzer-security.insecureAPI.strcpy]
2426 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2426)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/server/server.c:2426:21: note: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119
2426 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
Suppressed 51 warnings (42 in non-user code, 9 NOLINT).
make[2]: *** [Makefile:8807: examples/server/tests_unit_test-server.o] Error 1
make[2]: *** Waiting for unfinished jobs....
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2426)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/server/server.c:2426:21: warning: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119 [clang-analyzer-security.insecureAPI.strcpy]
2426 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2426)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/server/server.c:2426:21: note: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119
2426 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
Suppressed 51 warnings (42 in non-user code, 9 NOLINT).
make[2]: *** [Makefile:8877: examples/server/testsuite_testsuite_test-server.o] Error 1
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2407)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/client/client.c:2407:21: warning: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119 [clang-analyzer-security.insecureAPI.strcpy]
2407 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2407)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/client/client.c:2407:21: note: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119
2407 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
Suppressed 51 warnings (42 in non-user code, 9 NOLINT).
make[2]: *** [Makefile:8793: examples/client/tests_unit_test-client.o] Error 1
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2407)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/client/client.c:2407:21: warning: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119 [clang-analyzer-security.insecureAPI.strcpy]
2407 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
90fcd95f9b (<marco@wolfssl.com> 2022-08-10 16:41:42 +0200 2407)                     strcpy(dtlsCID, myoptarg);
/tmp/tmp.4346_11739/wolfssl_test_workdir.23721/wolfssl/examples/client/client.c:2407:21: note: Call to function 'strcpy' is insecure as it does not provide bounding of the memory buffer. Replace unbounded copy functions with analogous functions that support length arguments such as 'strlcpy'. CWE-119
2407 |                     strcpy(dtlsCID, myoptarg);
|                     ^~~~~~
Suppressed 51 warnings (42 in non-user code, 9 NOLINT).
make[2]: *** [Makefile:8835: examples/client/testsuite_testsuite_test-client.o] Error 1
make[1]: *** [Makefile:9063: all-recursive] Error 1
make: *** [Makefile:5432: all] Error 2
   real 1m35.559s  user 8m7.245s  sys 0m12.259s
    clang-tidy-all-sp-all fail_build
    failed config: '--srcdir' '.' '--disable-jobserver' '--enable-option-checking=fatal' '--enable-all' '--enable-testcert' '--enable-acert' '--enable-dtls13' '--enable-dtls-mtu' '--enable-dtls-frag-ch' '--enable-dtlscid' '--enable-quic' '--with-sys-crypto-policy' '--enable-sp-math-all' 'CPPFLAGS=-DWOLFSSL_SP_INT_NEGATIVE -DKEEP_OUR_CERT -DKEEP_PEER_CERT -DWOLFSSL_ALT_NAMES -DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK -pedantic -DSP_ALLOC -DWOLFSSL_CLANG_TIDY' 'CC=/home/douzzer/com/wolfssl/src/testing/git-hooks/clang-tidy-builder.sh' 'CLANG=/usr/lib/llvm/20/bin/clang-20' 'CFLAGS=-Wunreachable-code-aggressive -Wthread-safety -Wloop-analysis -Wenum-compare-conditional -fcolor-diagnostics -fcomplete-member-pointers -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wno-language-extension-token -g -fdebug-types-section -Wunreachable-code-break -Wunreachable-code-return -Wimplicit-fallthrough'
    BUILD_ENV: 'CLANG_TIDY=/usr/lib/llvm/20/bin/clang-tidy' 'CLANG=/usr/lib/llvm/20/bin/clang-20' 'CLANG_TIDY_EXTRA_ARGS=--use-color=1 -line-filter=[{"name":"asn1.h","lines":[[1,166]]},{"name":".c"},{"name":".h"},{"name":".s"},{"name":".S"},{"name":".i"}]' 'CLANG_OVERRIDE_CFLAGS=' 'FAIL_BUILD_CODENAME=fail_analytic_build' 'MAX_FIPS_CODE_SZ=10000000'
```
@rizlik may want to look at both of these.

julek-wolfssl committed:
Increase buffer to make room for \0

julek-wolfssl committed:
strcpy -> strncpy

julek-wolfssl commented:
@douzzer Learned to count and fixed the overflow. Fixed the clang-tidy errors too.

douzzer committed:
examples/client/client.c and examples/client/client.c: use XSTRLCPY() to assure proper null termination.

JacobBarthelmeh commented on .github/workflows/socat.yml:
```diff
@@ -78,4 +78,4 @@ jobs:
           export LD_LIBRARY_PATH=$GITHUB_WORKSPACE/build-dir/lib:$LD_LIBRARY_PATH
           export SHELL=/bin/bash
-          SOCAT=$GITHUB_WORKSPACE/socat-1.8.0.0/socat ./test.sh -t 0.5 --expect-fail 36,64,146,214,216,217,309,310,386,399,402,403,459,460,467,468,478,492,528,530
+          SOCAT=$GITHUB_WORKSPACE/socat-1.8.0.0/socat ./test.sh -t 0.5 --expect-fail 36,64,146,214,216,217,309,310,386,399,402,403,459,460,467,468,475,478,492,528,530
```
Why are we expecting more failures with socat unit tests now?

julek-wolfssl replied:
> socat.yml: in test 475, the test would send a SIGTERM after 3 seconds. We would continue to ignore this signal and continue to call `recvfrom`. Instead we should error out and give control back to the user.

Without this PR we would actually error out on `sendto`  with `EINVAL` due to passing in a bad peer parameter. With the peer handling refactoring we would no longer error out on `sendto` and be stuck in a `recvfrom` loop because we were ignoring the `EINTR` return. Now we return an error to the user.

JacobBarthelmeh replied:
Sorry, it's not clear to me. Is it that this PR fixed handling `EINVAL` which was causing the test to succeed previously but in an unexpected way? Where it should have been waiting and returning back on a `SIGTERM` signal? And now with this change we are ignoring a different signal that should be handled for the socat test?

julek-wolfssl replied:
Close. With the peer refactor we no longer trigger an invalid value error. When socat tries to kill the process it causes recvfrom to return with EINTR. Previously we would ignore this error and stall in an emdless recvfrom loop. Now I've changed the behaviour to return an error when the socket returns EINTR. Now we no longer stall but the socat test still fails. The test is trying to set the timeout directly on the socket. Our dtls code ignores this and overrides the timeout. I don't conaider this a failure so I've set the test to expect failure.

douzzer committed:
src/ssl.c: gate wolfSSL_dtls_set_pending_peer() on !defined(WOLFSSL_NO_SOCK), not just defined(WOLFSSL_DTLS_CID).

tests/api.c: in test_dtls12_basic_connection_id(), omit chacha20 suites if defined(HAVE_FIPS), and fix gate on DHE-PSK-NULL-SHA256.
